packages/api/src/globalContext.ts

```diff
@@ -1,5 +1,48 @@
+/* eslint-disable react-hooks/rules-of-hooks */
+
+// AWS Lambda run each request in a new process,
+// a process is not reused until a request is completed.
+//
+// Which means that each `global.context` is scoped to the lifetime of each request.
+
+// However when not in AWS Lambda, NodeJS is single-threaded, you must use the
+// per-request global context, otherwise you risk a race-condition
+// where one request overwrites another's global context.
+//
+// Alternatively only use the local `context` in a graphql resolver.
+
+import { AsyncLocalStorage } from 'async_hooks'
+
 export interface GlobalContext {
-  [key: string]: any
+  [key: string]: unknown
+}
+
+let GLOBAL_CONTEXT: GlobalContext = {}
+let PER_REQUEST_CONTEXT:
+  | undefined
+  | AsyncLocalStorage<Map<string, GlobalContext>> = undefined
+
+export const usePerRequestContext = () =>
+  process.env.SAFE_GLOBAL_CONTEXT !== '1'
```

Comment:
Should we set this to `'1'` in the netlify and vercel deploy setup scripts?

Reply:
Yeah - because there is a performance impact, but I'm not 100% certain that I want to trust Netlify and Vercel. So I would like to make this the default everywhere, and give people the ability to opt-out with the understand that it could be a problem.